### PR TITLE
Closes #32 Broken Smart Table Encoding

### LINKED ISSUE #32: Failing random smart table test
Not sure if this is something I broke with all the changes or whether it was already there, but in case it's easier for you to debug than me: `./build/smart_table_random 928496804` will fail for me on the 14th iteration.

### PR DESCRIPTION
**Pull request by mmcilree**

### CODE REVIEW

File: examples/smart_table_random/smart_table_random.cc [hunk 1]
```diff
@@ -139,23 +139,23 @@ auto test_smart_table(const int & n, mt19937 & rng)
                     }
                     shuffle(random_set.begin(), random_set.end(), rng);
                     int how_many = rand1_to_n(rng);
-                    // string_rep << "x[" << index_of(copy_x[split_points[i]], x) << "] ";
-                    // string_rep << constraint_type_str(constraint_type);
-                    // string_rep << " {";
+                    if (make_string_rep) string_rep << "x[" << index_of(copy_x[split_points[i]], x) << "] ";
```
Comment (ciaranm) [inline examples/smart_table_random/smart_table_random.cc:142]: What's wrong with an `if () {` block?

File: examples/smart_table_random/smart_table_random.cc [hunk 2]
```diff
@@ -224,46 +224,51 @@ auto test_smart_table(const int & n, mt19937 & rng)
...
-    solve_with(p,
+    auto stats = solve_with(p,
         SolveCallbacks{
-            .solution = [&](const CurrentState &) -> bool {
-                //                        cout << "x = [ ";
-                //                        for (const auto & var : x) {
-                //                            cout << s(var) << " ";
-                //                        }
-                //                        cout << "]" << endl;
+            .solution = [&](const CurrentState & s) -> bool {
```
Comment (ciaranm) [inline examples/smart_table_random/smart_table_random.cc:234]: Hrm, isn't this going to warn for unused s?

File: gcs/constraints/smart_table.cc [hunk 3]
```diff
@@ -733,9 +733,11 @@ auto SmartTable::install(Propagators & propagators, State & initial_state) && ->
 
                         auto flag = unary_set_entry.constraint_type == SmartEntryConstraint::In ? propagators.create_proof_flag("inset") : propagators.create_proof_flag("notinset");
 
-                        propagators.define(initial_state, move(set_value_sum) >= Integer{static_cast<long long>(set_value_sum.terms.size())},
+                        auto set_rhs = Integer{static_cast<long long>(set_value_sum.terms.size())};
```
Comment (ciaranm) [inline gcs/constraints/smart_table.cc:736]: Oh. That would do it... I wonder whether there's a linting tool or something that can catch that kind of thing.

**Review by ciaranm — APPROVED** [review]

File: examples/smart_table_random/smart_table_random.cc (hunk 1 above)
Comment (mmcilree) [reply examples/smart_table_random/smart_table_random.cc:142]: Whoops. Find and replace... 
Will remedy

File: gcs/constraints/smart_table.cc (hunk 3 above)
Comment (mmcilree) [reply gcs/constraints/smart_table.cc:736]: To be fair my IDE was flagging it, I had just overlooked / blanked the warning. Lesson learned

Comment (ciaranm) [reply gcs/constraints/smart_table.cc:736]: Ah, GCC doesn't seem to warn for this yet.

**Pull request merged**